nexus/src/db/datastore.rs

```diff
@@ -1011,24 +1182,66 @@ impl DataStore {
         &self,
         opctx: &OpContext,
         disk_authz: authz::ProjectChild,
+    ) -> DeleteResult {
+        let disk_id = disk_authz.id();
+        opctx.authorize(authz::Action::Delete, disk_authz).await?;
+        self.project_delete_disk_internal(
+            disk_id,
+            self.pool_authorized(opctx).await?,
+        )
+        .await
+    }
+
+    // TODO: Delete me (this function, not the disk!), ensure all datastore
+    // access is auth-checked.
+    //
+    // Here's the deal: We have auth checks on access to the database - at the
+    // time of writing this comment, only a subset of access is protected, and
+    // "Delete Disk" is actually one of the first targets of this auth check.
+    //
+    // However, there are contexts where we want to delete disks *outside* of
+    // calling the HTTP API-layer "delete disk" endpoint. As one example, during
+    // the "undo" part of the disk creation saga, we want to allow users to
+    // delete the disk they (partially) created.
+    //
+    // This gets a little tricky mapping back to user permissions - a user
+    // SHOULD be able to create a disk with the "create" permission, without the
+    // "delete" permission. To still make the call internally, we'd basically
+    // need to manufacture a token that identifies the ability to "create a
+    // disk, or delete a very specific disk with ID = ...".
+    pub async fn project_delete_disk_no_auth(
+        &self,
+        disk_id: &Uuid,
+    ) -> DeleteResult {
+        self.project_delete_disk_internal(disk_id, self.pool()).await
+    }
+
+    async fn project_delete_disk_internal(
```

Comment:
If I'm right that you can remove `Datastore::project_delete_disk()`, then you can also combine `project_delete_disk_no_auth` and `project_delete_disk_internal`, since there will be only one caller.